### ISSUE TITLE
[Flamingo] Flamingo model listeners cause UI component leaks

### ISSUE THREAD

**Issue by gLuEcKsKrAuT** [kirill-grouchnikov/radiance#283]
### Version of Radiance (current development is 3.5-SNAPSHOT)
3.0.0
### Sub-project (Neon, Trident, Substance, Flamingo, ...)
BasicCheckRibbon
### Version of Java (current minimum is 9)
9 (Open-JDK)
### Version of OS
Win 10 (1909)
### The issue you're experiencing (expected vs actual, screenshot, stack trace etc)
After resizing the BasicCheckRibbon frame and opening serveral popups at the paste button,
i watched some references with visual vm after a forced garbage collect.
* JRibbonTaskToggleButton
![ref-tasktoggle](https://user-images.githubusercontent.com/43807827/84376422-6a70f380-abe1-11ea-834b-dd2879038aa8.png)
    * expected arround 4 and not 70 instances
* JCommandPopupMenu (after 42 clicks)
![ref-popup](https://user-images.githubusercontent.com/43807827/84376926-23cfc900-abe2-11ea-8820-4d2935c5c126.png)
    * expected 0 instances and not 42

[visual-vm-heapdump-1591871903418.zip](https://github.com/kirill-grouchnikov/radiance/files/4764303/heapdump-1591871903418.zip)

**Comment by kirill-grouchnikov** [comment]
There's a lot of strongly-referenced listeners wired between the models and the UI class instances that represent those models on screen. It requires non-trivial amount of plumbing to convert those to weak references, so it will happen in 3.5-SNAPSHOT and not be done as part of a bug-fix for 3.0.0 (and in any case, that ship has sailed).

**Comment by gLuEcKsKrAuT** [comment]
seems to work for the mentioned that cases, 
thx for the fix.

if it helps, for my stuff i use something like this,
to not implement farms of specialized listener-list's or weak wrappers.

free for use of course.

```java
import java.lang.ref.Reference;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;

/**
 * Support for handling listeners. <br/>
 * Functionality is similiar to {@link javax.swing.event.EventListenerList}. <br/>
 * There is also a option to refer given listeners weak.
 *
 * @author gLuEcKsKrAuT
 * @see IWeakListener
 * @since 13.06.2020
 */
public final class EventListenerMap
{
    private Map<Class<?>, List<Object>> listenerMap;

    /**
     * Add given listener. <br/>
     * Similar to {@code EventListenerMap#add(type, listener, false)}.
     *
     * @param type     Type of listener(s)
     * @param listener Listener
     * @param <L>      Listener type
     *
     * @see #add(Class, Object, boolean)
     */
    public <L> void add (Class<L> type, L listener) {
        add(type, listener, false);
    }

    /**
     * Add given listener. <br/>
     * The listener will be referred weak if: <br/>
     * - {@code weak = true} or <br/>
     * - listener is instance of {@link IWeakListener}
     *
     * @param type     Type of listener(s)
     * @param listener Listener
     * @param weak     Refer given listener weak
     * @param <L>      Listener type
     */
    public <L> void add (Class<L> type, L listener, boolean weak) {
        getListenerMap(true)
            .computeIfAbsent(type, EventListenerMap::_createListenerCol)
            .add(_createListener(listener, requireWeak(listener, weak)));
    }

    /**
     * Publish given event to listener(s) of given type.
     * <pre>{@code
     *      EventListenerMap lm = ... ;
     *
     *      lm.fire(
     *          ChangeListener.class,
     *          new ChangeEvent(...),
     *          ChangeListener::stateChanged
     *      );
     * }</pre>
     *
     * @param type    Type of listener(s)
     * @param event   Event to publish
     * @param handler Handle that dispatch the event to a listener
     * @param <L>     Listener type
     * @param <E>     Event type
     */
    public <L, E> void fire (Class<L> type, E event, BiConsumer<L, E> handler) {
        for (L listener : getListeners(type)) {
            handler.accept(listener, event);
        }
    }

    /**
     * Get listener(s) of given type.
     * <pre>{@code
     *      EventListenerMap lm = ... ;
     *      ChangeEvent ce = ... ;
     *
     *      for (ChangeListener l : lm.getListeners(ChangeListener.class)) {
     *          l.stateChanged(e);
     *      }
     * }</pre>
     *
     * @param type Type of listener(s)
     * @param <L>  Listener type
     *
     * @return {@code List<T>} of listener(s) or {@code empty()}
     */
    public <L> List<L> getListeners (Class<L> type) {
        Map<Class<?>, List<Object>> listenerMap = getListenerMap(false);

        if (listenerMap == null) {
            return Collections.emptyList();
        }

        List<Object> listeners = listenerMap.get(type);

        if (listeners == null) {
            return Collections.emptyList();
        }

        List<L> result = new ArrayList<>();

        for (int i = listeners.size() - 1; i >= 0; i--) {
            Object listener = listeners.get(i);
            if (listener instanceof Reference) {
                Reference<?> ref               = (Reference<?>) listener;
                Object       listenerUnwrapped = ref.get();
                if (listenerUnwrapped != null) {
                    result.add((type.cast(listenerUnwrapped)));
                }
                else {
                    listeners.remove(listener);
                }
            }
            else {
                result.add(type.cast(listener));
            }
        }

        Collections.reverse(result);

        return result;
    }

    private Map<Class<?>, List<Object>> getListenerMap (boolean create) {
        return listenerMap != null ? listenerMap : create ? (listenerMap = _createListenerMap()) : Collections.emptyMap();
    }

    private static <L> Object _createListener (L listener, boolean requireWeak) {
        return requireWeak ? listener instanceof WeakReference ? listener : new WeakReference<>(listener) : listener;
    }

    private static <T> List<T> _createListenerCol (Class<?> type) {
        return new ArrayList<>();
    }

    private static <K, V> Map<K, V> _createListenerMap () {
        return new HashMap<>();
    }

    private static boolean requireWeak (Object listener, boolean weak) {
        return weak || listener instanceof IWeakListener;
    }

    /** Marker for listener that should be referred only weak */
    public interface IWeakListener {}
}
```
